File: src/agents/agent.py
"""Agent (Adventurer) for the dungeon."""
from enum import IntEnum
from ..core.grid import Grid
from ..core.tiles import TileType, is_passable, get_reward


class Action(IntEnum):
    """Actions the agent can take."""
    UP = 0
    DOWN = 1
    LEFT = 2
    RIGHT = 3


# Direction vectors for each action (dx, dy)
ACTION_DELTAS = {
    Action.UP: (0, -1),
    Action.DOWN: (0, 1),
    Action.LEFT: (-1, 0),
    Action.RIGHT: (1, 0),
}


class Agent:
    """An agent that navigates the dungeon."""

    def __init__(self, x: int, y: int, hp: int = 100, max_hp: int = 100):
        """Initialize the agent.

        Args:
            x: Starting x position
            y: Starting y position
            hp: Starting health points
            max_hp: Maximum health points
        """
        self.x = x
        self.y = y
        self.hp = hp
        self.max_hp = max_hp
        self.total_reward = 0.0

    @property
    def position(self) -> tuple[int, int]:
        """Get current position."""
        return (self.x, self.y)

    @property
    def is_alive(self) -> bool:
        """Check if the agent is alive."""
        return self.hp > 0

    def get_next_position(self, action: Action) -> tuple[int, int]:
        """Calculate the next position for a given action.

        Args:
            action: The action to take

        Returns:
            The (x, y) position after the action
        """
        dx, dy = ACTION_DELTAS[action]
        return (self.x + dx, self.y + dy)

    def can_move(self, action: Action, grid: Grid) -> bool:
        """Check if a move is valid.

        Args:
            action: The action to take
            grid: The grid to check against

        Returns:
            True if the move is valid
        """
        next_x, next_y = self.get_next_position(action)

        # Check bounds
        if not grid.is_valid_position(next_x, next_y):
            return False

        # Check if tile is passable
        tile = grid.get_tile(next_x, next_y)
        return is_passable(tile)

    def move(self, action: Action, grid: Grid) -> tuple[float, bool, bool]:
        """Execute a move action.

        Args:
            action: The action to take
            grid: The grid to move on

        Returns:
            Tuple of (reward, done, success)
            - reward: The reward for this step
            - done: True if episode ended (goal reached or died)
            - success: True if move was successful
        """
        step_reward = -0.1  # Small penalty for each step

        if not self.can_move(action, grid):
            # Wall bump
            self.total_reward += step_reward - 1
            return step_reward - 1, False, False

        # Execute move
        self.x, self.y = self.get_next_position(action)

        # Get tile at new position
        tile = grid.get_tile(self.x, self.y)
        tile_reward = get_reward(tile)

        # Apply tile effects
        done = False
        if tile == TileType.GOAL:
            done = True
        elif tile == TileType.TRAP:
            self.hp -= 10
            if self.hp <= 0:
                done = True
        elif tile == TileType.HEAL:
            self.hp = min(self.hp + 10, self.max_hp)

        total_step_reward = step_reward + tile_reward
        self.total_reward += total_step_reward

        return total_step_reward, done, True

    def reset(self, x: int, y: int):
        """Reset the agent to a new position.

        Args:
            x: New x position
            y: New y position
        """
        self.x = x
        self.y = y
        self.hp = self.max_hp
        self.total_reward = 0.0


def random_action() -> Action:
    """Return a random action."""
    import random
    return Action(random.randint(0, 3))

File: web/js/game/agent.js
/**
 * Agent (Adventurer) for the dungeon
 */

import { TileType, isPassable, getReward, isLethal, getTileDamage } from './tiles.js';

export const Action = {
    UP: 0,
    DOWN: 1,
    LEFT: 2,
    RIGHT: 3
};

export const ActionNames = ['UP', 'DOWN', 'LEFT', 'RIGHT'];

export const ActionDeltas = {
    [Action.UP]: { dx: 0, dy: -1 },
    [Action.DOWN]: { dx: 0, dy: 1 },
    [Action.LEFT]: { dx: -1, dy: 0 },
    [Action.RIGHT]: { dx: 1, dy: 0 }
};

export class Agent {
    constructor(x, y, hp = 100, maxHp = 100) {
        this.x = x;
        this.y = y;
        this.hp = hp;
        this.maxHp = maxHp;
        this.totalReward = 0;
        this.turnCount = 0;
        this.actionHistory = [];
        this.visitHistory = new Map(); // "x,y" -> last visit turn
        // T2B-2: optional ModifierSet (slippery / heavy_fog). Set by main.js.
        this.modifierSet = null;
        // T2B-2: heavy_fog reduces this; default 5 reproduces legacy decay 1.0/0.8/0.6/0.4/0.2.
        this.visibilityRange = 5;
        this._recordVisit();
    }

    _recordVisit() {
        const key = `${this.x},${this.y}`;
        this.visitHistory.set(key, this.turnCount);
    }

    getVisibility(x, y) {
        // Returns opacity (0-1) based on how recently the cell was visited.
        const key = `${x},${y}`;
        if (!this.visitHistory.has(key)) {
            return 0; // Never visited - fog
        }
        const lastVisit = this.visitHistory.get(key);
        const turnsSince = this.turnCount - lastVisit;
        const range = this.visibilityRange || 5;
        if (turnsSince >= range) return 0;
        // Linear decay; at range=5 yields 1.0/0.8/0.6/0.4/0.2 (legacy values).
        return Math.max(0, 1 - turnsSince / range);
    }

    get position() {
        return { x: this.x, y: this.y };
    }

    get isAlive() {
        return this.hp > 0;
    }

    getNextPosition(action) {
        const delta = ActionDeltas[action];
        return {
            x: this.x + delta.dx,
            y: this.y + delta.dy
        };
    }

    canMove(action, grid) {
        const next = this.getNextPosition(action);

        if (!grid.isValidPosition(next.x, next.y)) {
            return false;
        }

        const tile = grid.getTile(next.x, next.y);
        return isPassable(tile);
    }

    /**
     * Apply stochastic transition.
     *   - grid.slippery (FrozenLake config-driven, Lv.26~28): 1/3 intended,
     *     1/3 perpendicular-left, 1/3 perpendicular-right (Math.random, legacy).
     *   - Modifier 'slippery' (T2B-2 runtime): 30% sideways via seeded RNG.
     *
     * D-2026-05-12-10: grid.slippery takes precedence — the modifier path only
     * runs when grid.slippery is false. We deliberately do NOT stack the two
     * slip models even if both are present (the FrozenLake dungeons are
     * already balanced around 2/3 slip).
     */
    _resolveAction(action, grid) {
        if (this.modifierSet && !grid.slippery) {
            const deflected = this.modifierSet.resolveMovement(action);
            if (deflected != null) return deflected;
        }
        if (!grid.slippery) return action;

        const r = Math.random();
        if (r < 1/3) {
            return action; // Intended direction
        } else if (r < 2/3) {
            // Perpendicular left (rotate -90°)
            // UP->LEFT, DOWN->RIGHT, LEFT->DOWN, RIGHT->UP
            return [Action.LEFT, Action.RIGHT, Action.DOWN, Action.UP][action];
        } else {
            // Perpendicular right (rotate +90°)
            // UP->RIGHT, DOWN->LEFT, LEFT->UP, RIGHT->DOWN
            return [Action.RIGHT, Action.LEFT, Action.UP, Action.DOWN][action];
        }
    }

    move(action, grid) {
        const stepReward = -0.1;  // Small penalty for each step

        // Stochastic transition: resolve actual action
        const resolvedAction = this._resolveAction(action, grid);
        this.actionHistory.push(resolvedAction);

        if (!this.canMove(resolvedAction, grid)) {
            // Wall bump (stay in place, still penalized)
            const penalty = stepReward - 1;
            this.totalReward += penalty;
            return { reward: penalty, done: false, success: false };
        }

        // Execute move with resolved action
        const next = this.getNextPosition(resolvedAction);
        this.x = next.x;
        this.y = next.y;
        this.turnCount++;
        this._recordVisit();

        // Get tile at new position
        const tile = grid.getTile(this.x, this.y);
        const tileReward = getReward(tile);

        // Apply tile effects
        let done = false;

        if (tile === TileType.GOAL) {
            done = true;
        } else if (tile === TileType.PIT) {
            // 즉사 함정
            this.hp = 0;
            done = true;
        } else if (tile === TileType.TRAP) {
            this.hp -= 10;
            if (this.hp <= 0) {
                this.hp = 0;
                done = true;
            }
        } else if (tile === TileType.HEAL) {
            this.hp = Math.min(this.hp + 10, this.maxHp);
        } else if (tile === TileType.GOLD) {
            // 골드는 보상만 (나중에 골드 시스템에서 처리)
        } else if (tile === TileType.MONSTER) {
            // 몬스터: 높은 데미지, 처치 후 사라짐
            const damage = getTileDamage(tile);
            this.hp -= damage;
            if (this.hp <= 0) {
                this.hp = 0;
                done = true;
            }
        }

        const totalStepReward = stepReward + tileReward;
        this.totalReward += totalStepReward;

        return { reward: totalStepReward, done, success: true, tile };
    }

    reset(x, y) {
        this.x = x;
        this.y = y;
        this.hp = this.maxHp;
        this.totalReward = 0;
        this.turnCount = 0;
        this.actionHistory = [];
        this.visitHistory.clear();
        // T2B-2: fail-safe defaults — callers (main.js) must re-apply modifiers
        // via _syncAgentModifiers if a daily ModifierSet should persist.
        this.modifierSet = null;
        this.visibilityRange = 5;
        this._recordVisit();
    }
}

export function randomAction() {
    return Math.floor(Math.random() * 4);
}
